### PR TITLE
Use HikariCP for connection pooling

### PR DESCRIPTION
**Pull request by ammaraskar**
Will probably result in way better performance than whatever this weird connection pool that's currently being used.

Caveats:

* Increases the jar size from ~200kb to ~1000kb (this shouldn't be too big a concern)

### CODE REVIEW

File: src/main/java/de/diddiz/util/MySQLConnectionPool.java [hunk 1]
```diff
@@ -1,437 +1,39 @@
...
+		ds.setPassword(password);
+
+		ds.addDataSourceProperty("useUnicode", "true");
+		ds.addDataSourceProperty("characterEncoding", "utf-8");
+		ds.addDataSourceProperty("rewriteBatchedStatements", "true");
+
+		ds.addDataSourceProperty("cachePrepStmts", "true");
+		ds.addDataSourceProperty("prepStmtCacheSize", "250");
+		ds.addDataSourceProperty("prepStmtCacheSqlLimit", "2048");
+		ds.addDataSourceProperty("useServerPrepStmts", "true");
```
Comment (frymaster) [inline src/main/java/de/diddiz/util/MySQLConnectionPool.java]: Apparently you can't use this and ``rewriteBatchedStatements`` at the same time according to the entry for the rewrite parameter in http://dev.mysql.com/doc/connector-j/en/connector-j-reference-configuration-properties.html

I'll do some benchmarking this weekend and check - the batch rewriting is what gave us most of the speed boost, so I'd rather keep that if it's a choice

Comment (ammaraskar) [inline src/main/java/de/diddiz/util/MySQLConnectionPool.java]: Good point, I think batch rewriting will probably end up on top but benchmarking is always good.

File: src/main/java/de/diddiz/util/MySQLConnectionPool.java [hunk 2]
```diff
@@ -1,437 +1,39 @@
...
-				}
-			}
-			final JDCConnection conn = new JDCConnection(DriverManager.getConnection(url, user, password));
-			conn.lease();
-			if (!conn.isValid()) {
-				conn.terminate();
-				throw new SQLException("Failed to validate a brand new connection");
-			}
-			connections.add(conn);
-			if (mb4) conn.createStatement().executeQuery("SET NAMES utf8mb4");
```
Comment (frymaster) [inline src/main/java/de/diddiz/util/MySQLConnectionPool.java]: See lines 94 and 95 of [LogBlock.java](https://github.com/LogBlock/LogBlock/blob/master/src/main/java/de/diddiz/LogBlock/LogBlock.java#L94-L95)

This is from commit 82b4ffc2a28eef2d45713bbf5726d7c31d44e2d4 which adds support for the full set of Unicode points.  I found I needed to explicitly do that query or, even if the database column supported it, it still wouldn't allow them

Not sure we'd set that

Comment (ammaraskar) [inline src/main/java/de/diddiz/util/MySQLConnectionPool.java]: Now returning connections with the SET NAMES query.

Comment (frymaster) [inline src/main/java/de/diddiz/util/MySQLConnectionPool.java]: Confirming that unfortunately yes, that was still needed for me :(

(I think if the server defaults to utf8mb4 it's not needed, so some people's tests may vary)

**Comment by frymaster** [comment]
As per [the docs](https://github.com/brettwooldridge/HikariCP#configuration-knobs-baby) HikariCP defaults to using a fixed-size pool so it doesn't have to spend the extra time creating a new connection at the time of demand.  That's sensible for almost all situations, however with LogBlock we've got 1 consumer thread, and a varying number of query threads - quite often none - some of which are running for a decent amount of time

As such I think it'd be worth setting `` minimumIdle `` to 1, so we don't instantly use all 10 db connections all the time, as some shared hosting only allows 10 DB connections from a single username *in total*, and they may have other MySQL-using plugins

**Comment by ammaraskar** [comment]
I've set it to 2, just in case we need an extra connection somewhere at any point, such as when we're looking up a player's UUID on join while the consumer is running.

**Comment by frymaster** [comment]
Been working fine over the weekend

**Pull request merged**